File: f3_fmod/NativeFmodEx/Src/NativeFmodDesigner/Native/CallbackManager.cpp
/**
 * 				NativeFmodEx Project
 *
 * Want to use FMOD Ex API (www.fmod.org) in the Java language ? NativeFmodEx is made for you.
 * Copyright © 2005-2008 Jérôme JOUVIE (Jouvieje)
 *
 * Created on 23 feb. 2005
 * @version file v1.4.4
 * @author Jérôme JOUVIE (Jouvieje)
 * 
 * 
 * WANT TO CONTACT ME ?
 * E-mail :
 * 		jerome.jouvie@gmail.com
 * My web sites :
 * 		http://jerome.jouvie.free.fr/
 * 
 * 
 * INTRODUCTION
 * FMOD Ex is an API (Application Programming Interface) that allow you to use music
 * and creating sound effects with a lot of sort of musics.
 * FMOD is at :
 * 		http://www.fmod.org/
 * The reason of this project is that FMOD Ex can't be used direcly with Java, so I've created
 * this project to do this.
 * 
 * 
 * GNU LESSER GENERAL PUBLIC LICENSE
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of the License,
 * or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA 
 */
#include "Utils.h"
#include "Pointer.h"
#include "fmod_event.h"
#include "fmod_event_net.h"
#include "fmod_event.hpp"
#include "malloc.h"
#include "CallbackManager.h"

jclass byteBufferClass = 0;
jclass getByteBufferClass(JNIEnv *java_env) {
	if(!byteBufferClass) {
		byteBufferClass = (jclass)java_env->NewGlobalRef(java_env->FindClass("java/nio/ByteBuffer"));
	}
	return byteBufferClass;
}

jclass caller = 0;
void connectCaller(JNIEnv *java_env) {
	caller = (jclass)java_env->NewGlobalRef(java_env->FindClass("org/jouvieje/FmodDesigner/Callbacks/CallbackBridge"));
	if(java_env->ExceptionCheck()) {
		java_env->ExceptionClear();
		caller = 0;
		ThrowException(java_env, InitException, "Connection to CallbackBridge fails.");
	}
}

jmethodID callbackId[2];
void connectCallbacks(JNIEnv *java_env) {
	static struct {
		const char *name;
		const char *signature;
	}callbacks[2] = {
		{"FMOD_EVENT_CALLBACK_BRIDGE", "(JIJLorg/jouvieje/FmodEx/Misc/Pointer;J)I"},
		{"FMOD_MUSIC_CALLBACK_BRIDGE", "(IJJJ)I"}
	};

	for(int i = 0; i < 2; i++) {
		callbackId[i] = java_env->GetStaticMethodID(caller, callbacks[i].name, callbacks[i].signature);
		if(java_env->ExceptionCheck()) {
			java_env->ExceptionClear();
			ThrowException(java_env, InitException, "Connection to a Callback fails.");
			return;
		}
	}
}

JavaVM *jvm;
void attachJavaVM(JNIEnv *java_env) {
	java_env->GetJavaVM(&jvm);
	connectCaller(java_env);
	if(caller) {
		connectCallbacks(java_env);
	}
}
bool acquire_jenv(JNIEnv **java_env) {
	if(jvm->GetEnv((void **)java_env, JNI_VERSION_1_4) != JNI_OK) {
		jvm->AttachCurrentThread((void **)java_env, 0);
		return true;
	}
	return false;
}
void leave_jenv(bool attached) {
	if(attached) {
		jvm->DetachCurrentThread();
	}
}

FMOD_RESULT F_CALLBACK FMOD_EVENT_CALLBACK_BRIDGE(FMOD_EVENT * event, FMOD_EVENT_CALLBACKTYPE type, void * param1, void * param2, void * userdata) {
	JNIEnv *java_env/* = 0*/;
	bool attached = acquire_jenv(&java_env);
	POINTER_TYPE jevent/* = 0*/;
	*(FMOD_EVENT **)&jevent = (FMOD_EVENT *)event;
	POINTER_TYPE jparam1/* = 0*/;
	*(void **)&jparam1 = param1;
	POINTER_TYPE param2NewPointerAddress/* = 0*/;
	*(void **)&param2NewPointerAddress = param2;
	jobject jparam2 = newPointer(java_env, param2NewPointerAddress);
	POINTER_TYPE juserdata/* = 0*/;
	*(void **)&juserdata = userdata;
	jint result_ = java_env->CallStaticIntMethod(caller, callbackId[0], (jlong)jevent, (jint)type, (jlong)jparam1, (jobject)jparam2, (jlong)juserdata);
	if(java_env->ExceptionCheck()) {
		java_env->Throw(java_env->ExceptionOccurred());
		if(java_env->ExceptionCheck()) {
			java_env->ExceptionDescribe();
			java_env->FatalError(FATAL_ERROR_MESSAGE);
		}
	}
	if(jparam2 && (type == FMOD_EVENT_CALLBACKTYPE_SOUNDDEF_CREATE)) {
		POINTER_TYPE jparam2Address = getPointerAddress(java_env, jparam2);
		if(jparam2Address) {
			*(FMOD::Sound **)param2 = *(FMOD::Sound **)&jparam2Address;
		}
	}
	leave_jenv(attached);
	return (FMOD_RESULT)result_;
}

FMOD_RESULT F_CALLBACK FMOD_MUSIC_CALLBACK_BRIDGE(FMOD_MUSIC_CALLBACKTYPE type, void * param1, void * param2, void * userdata) {
	JNIEnv *java_env/* = 0*/;
	bool attached = acquire_jenv(&java_env);
	POINTER_TYPE jparam1/* = 0*/;
	*(void **)&jparam1 = param1;
	POINTER_TYPE jparam2/* = 0*/;
	*(void **)&jparam2 = param2;
	POINTER_TYPE juserdata/* = 0*/;
	*(void **)&juserdata = userdata;
	jint result_ = java_env->CallStaticIntMethod(caller, callbackId[1], (jint)type, (jlong)jparam1, (jlong)jparam2, (jlong)juserdata);
	if(java_env->ExceptionCheck()) {
		java_env->Throw(java_env->ExceptionOccurred());
		if(java_env->ExceptionCheck()) {
			java_env->ExceptionDescribe();
			java_env->FatalError(FATAL_ERROR_MESSAGE);
		}
	}
	leave_jenv(attached);
	return (FMOD_RESULT)result_;
}

File: f3_fmod/NativeFmodEx/Src/NativeFmodEx/Java/org/jouvieje/FmodEx/Callbacks/CallbackManager.java
/**
 * 							NativeFmodEx Project
 *
 * Do you want to use FMOD Ex API (www.fmod.org) with the Java language ? I've created NativeFmodEx for you.
 * Copyright © 2005 Jérôme JOUVIE (Jouvieje)
 *
 * Created on 23 feb. 2005
 * @version file v1.0.0
 * @author Jérôme JOUVIE (Jouvieje)
 * 
 * 
 * WANT TO CONTACT ME ?
 * E-mail :
 * 		jerome.jouvie@gmail.com
 * My web sites :
 * 		http://jerome.jouvie.free.fr/
 * 		http://topresult.tomato.co.uk/~jerome/
 * 
 * 
 * INTRODUCTION
 * FMOD Ex is an API (Application Programming Interface) that allow you to use music
 * and creating sound effects with a lot of sort of musics.
 * FMOD is at :
 * 		http://www.fmod.org/
 * The reason of this project is that FMOD Ex can't be used direcly with Java, so I've created
 * this project to do this.
 * 
 * 
 * GNU LESSER GENERAL PUBLIC LICENSE
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of the License,
 * or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA 
 */

package org.jouvieje.FmodEx.Callbacks;

import java.util.Enumeration;
import java.util.Hashtable;

import org.jouvieje.FmodEx.Init;
import org.jouvieje.FmodEx.Defines.INIT_MODES;
import org.jouvieje.FmodEx.Exceptions.CallbackException;

//TODO Make this class invisible

/**
 * <b><u>WARNING</u></b><br>
 * DO NOT USE THIS CLASS !!!
 */
public class CallbackManager
{
	public static boolean DEBUG_MODE = false;
	
	//Variable used to know if INIT_FMOD_EX_MINIMUM was used to initialized FMOD_EX
	protected static boolean useCallbacks = false;
	protected final static String ERROR_MESSAGE = "A callback was used with Init.loadLibraries(INIT_FMOD_***_MINIMUM) !";
	
	/*
	 * Key   : owner    (Long)
	 * Value : Callback (Object)
	 */
	protected static Hashtable[] callbacksTable = null;
	//Store the last callback added
	protected static Object[] lastCallbacksAdded = null;
	/*
	 * Key   : object (Long)
	 * Value : owner (Long)
	 */
	protected static Hashtable ownersTable = null;
	
	static
	{
		if(Init.getInitMode() == INIT_MODES.INIT_FMOD_EX)
		{
			callbacksTable = new Hashtable[CallbackBridge.NB_CALLBACKS];
			lastCallbacksAdded = new Object[CallbackBridge.NB_CALLBACKS];
			ownersTable = new Hashtable();
			for(int i = 0; i < callbacksTable.length; i++)
			{
				callbacksTable[i] = new Hashtable();
				lastCallbacksAdded[i] = null;
			}
			useCallbacks = true;
		}
	}
	
	public static Object getCallback(int type)
	{
		return getCallback(type, 0, false);
	}
	public static Object getCallback(int type, long object, boolean autoAttach)
	{
		if(!useCallbacks)
			throw new RuntimeException(ERROR_MESSAGE);
		
		Long owner = (object == 0) ? new Long(0) : (Long)ownersTable.get(new Long(object));
		if(owner != null)
		{
			Object callback = callbacksTable[type].get(owner);
			if(callback == null)
			{
				if(object == 0)
				{
					if(DEBUG_MODE)
						printDebug("Owner not found (type="+type+" object="+object+")! Try to use last Callbacks added");
					callback = lastCallbacksAdded[type];
				}
				if(callback == null)
					throw new CallbackException("A callback may not be implemented for the moment. Contact NativeFmodEx support to know more about this.");
			}
			if(DEBUG_MODE)
				printDebug("Get Callback type="+type+"("+getCallbackName(type)+") owner="+owner);
			return callback;
		}
		else
		{
			if(DEBUG_MODE)
				printDebug("Owner not found (type="+type+" object="+object+")! Use last Callbacks.");
			Object callback = lastCallbacksAdded[type];
			if(callback == null)
			{
				throw new CallbackException("A callback may not be implemented for the moment. Contact NativeFmodEx support to know more about this.");
			}
			if(autoAttach)
			{
				Enumeration values = callbacksTable[type].elements();
				Enumeration keys = callbacksTable[type].keys();
				while(values.hasMoreElements())
				{
					Object currentValue = values.nextElement();
					Long currentKey = (Long)keys.nextElement();
					if(currentValue == callback)
					{
						owner = currentKey;
						addOwner(owner.longValue(), object);
						break;
					}
				}
			}
			if(DEBUG_MODE)
				printDebug("Get Callback type="+type+"("+getCallbackName(type)+") owner="+owner);
			return callback;
		}
	}
	
	public static void addTmpCallback(int type, Object callback)
	{
		if(!useCallbacks)
		{
			if(callback != null)
				throw new RuntimeException(ERROR_MESSAGE);
			return;
		}
		
		if(callback != null)
		{
			lastCallbacksAdded[type] = callback;

			if(DEBUG_MODE)
				printDebug("Add Temporary Callback type="+type);
		}
	}
	public static void addCallback(int type, Object callback, long owner)
	{
		if(!useCallbacks)
		{
			if(callback != null)
				throw new RuntimeException(ERROR_MESSAGE);
			return;
		}
		
		if(callbacksTable[type].remove(new Long(owner)) != null)
		{
			lastCallbacksAdded[type] = null;
			if(DEBUG_MODE)
				printDebug("Removing Callback type="+type+" owner="+owner);
		}
		if(callback != null)
		{
			callbacksTable[type].put(new Long(owner), callback);
			lastCallbacksAdded[type] = callback;
			
			if(DEBUG_MODE)
				printDebug("Add Callback type="+type+" owner="+owner);
		}
	}
	
	public static void addOwner(long owner, long object)
	{
		if(!useCallbacks)
			return;
		
		Object o = ownersTable.remove(new Long(object));
		if(o != null)
		{
			if(DEBUG_MODE)
				printDebug("Removing Owner owner="+(o != null ? ((Long)o).longValue() : 0)+" object="+object);
		}
		if(owner != 0)
		{
			ownersTable.put(new Long(object), new Long(owner));
			
			if(DEBUG_MODE)
				printDebug("Add Owner owner="+owner+" object="+object);
		}
	}
	
	public static String getCallbackName(int type)
	{
		return CallbackBridge.getCallbackName(type);
	}
	
						/* For DEBUG mode*/
	
	private static void printDebug(String message)
	{
		java.lang.System.out.println("CALLBACK MANAGER : "+message);
	}
}